src/main/java/org/commonjava/indy/service/archive/controller/ArchiveController.java

```diff
@@ -0,0 +1,256 @@
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+@ApplicationScoped
+public class ArchiveController {
+
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    @Inject
+    PreSeedConfig preSeedConfig;
+
+    @Inject
+    ObjectMapper objectMapper;
+
+    private ExecutorService executorService;
+
+    private CloseableHttpClient client;
+
+    private String contentDir;
+
+    private String archiveDir;
+
+    private final String CONTENT_DIR = "/content";
+
+    private final String ARCHIVE_DIR = "/archive";
+
+    @PostConstruct
+    public void init()
+    {
+        int threads = 4 * Runtime.getRuntime().availableProcessors();
+        executorService = Executors.newFixedThreadPool( threads, ( final Runnable r ) -> {
+            final Thread t = new Thread( r );
+            t.setDaemon( true );
+            return t;
+        } );
+
+        final PoolingHttpClientConnectionManager ccm = new PoolingHttpClientConnectionManager();
+        ccm.setMaxTotal( 500 );
+
+        RequestConfig rc = RequestConfig.custom().build();
+        final HttpClientBuilder builder = HttpClients.custom().setConnectionManager( ccm ).setDefaultRequestConfig( rc );
+        client = builder.build();
+
+        String storeDir = preSeedConfig.storageDir.orElse( "data" );
+        contentDir = String.format( "%s%s", storeDir, CONTENT_DIR );
+        archiveDir = String.format( "%s%s", storeDir, ARCHIVE_DIR );
+    }
+
+    @PreDestroy
+    public void destroy()
+    {
+        IOUtils.closeQuietly( client, null );
+    }
+
+    public void downloadArtifacts( final Map<String, String> downloadPaths, final HistoricalContentDTO content )
+    {
+        BasicCookieStore cookieStore = new BasicCookieStore();
+        ExecutorCompletionService executor = new ExecutorCompletionService<>( executorService );
+
+        String contentBuildDir = String.format( "%s/%s", contentDir, content.getBuildConfigId() );
+        File dir = new File( contentBuildDir );
+        dir.deleteOnExit();
+
+        fileTrackedContent( contentBuildDir, content );
+
+        for ( String path : downloadPaths.keySet() )
+        {
+            String filePath = downloadPaths.get( path );
+            executor.submit( download( contentBuildDir, path, filePath, cookieStore ) );
```

Comment:
If I understand it correctly, these downloading will happen async, but in your resource you did not give any control on the completion,so this will cause the data corruption, which means some un-finished downloading will happen when you start doing the archive.

Reply:
Yes, I've watched